server/src/main/java/org/elasticsearch/features/FeatureService.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.features;
+
+import org.elasticsearch.Version;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+
+import static java.util.Map.entry;
+
+/**
+ * Service responsible for registering features this node should publish to other nodes
+ */
+public class FeatureService {
+
+    /**
+     * The set of features that are implied by historical node versions.
+     */
+    private static final Map<Version, Set<String>> HISTORICAL_FEATURES = Map.ofEntries(
+        entry(Version.V_8_10_0, Set.of("elastic-connectors.templates")),
+        entry(Version.V_8_11_0, Set.of("esql.stats_node"))
```

Comment:
This whole thing is a massive information leak from xpack. Also these plugins might not be installed!

Reply:
Yes, that's not ideal. Maybe we can replicate the same schema in each of the plugins, building historical feature maps in each of them and registering them as regular features via PluginService - there should be no difference. It's more effort, but it will be more precise (if a plugin is not installed the feature is not there) and contain information better.

Reply:
We could have a `FeatureProvidingPlugin` interface that could return a map of legacy Version-linked features and a set of new features. But once we're defining features in plugins, I think we need to add some kind of name spacing so that the string identifiers can't clash.

Node startup could refuse to proceed if two plugins try to provide the same feature.

Reply:
Name spacing was also on my mind - I had it in my own take(s) at Features previously. One nice side effect is to be able to quickly identify a feature (set) owner. I don't know if we need it now, but it might be one of those things it would be difficult to introduce later.

Reply:
That's a general thing though - node startup fails if anything provides two features with the same name.

I like the idea of namespacing plugin features, but is there any use case for moving features between plugins? That would cause lots of problems, with no way to affect the namespacing.

Reply:
I've added historical feature registration. Because of its use in `DiscoveryNode`, it needs to be static - which is not great, but I don't think there's another way to do it

Reply:
I understand your weariness: you have plugin instance register -> static registration method -> static historic feature set <- static read method <- discoverynode instance.

Timing works out correct but you dependent on it, without a good way of enforcing it.

I would like to revisit some sort of "static initialization is registration".
Would you think something like 
`static final NodeFeature ESQL_STATS_FEATURE = FeatureService.registerHistoricalFeature("esql.stats_node",  V_8_10_0);` could work?

It would create the `NodeFeature` with the correct era and call the current `FeatureService.registerHistoricalFeature` implementation (or just replace it).

For consistency, we could have 
`static final NodeFeature ESQL_STATS_FEATURE = FeatureService.registerFeature("esql.some_other_feature",  8);` too; that also creates the NodeFeature, and then proceed with the current `registerFeature` logic.

This way you give up the check on the pluginEra / declared era. But let me ask: is it important? 
My understanding is that in the current status we can just use the node version era, because plugins will have the same version; maybe in the future we will have relax it, but do we expect a cluster to load plugins from a different major? 
And even if we end up registering a feature for a past era, would it be bad?

I am pretty sure I could be missing something important thought :)

Reply:
I don't think the whole `FeatureService` class can be static, as we need to be able to handle multiple `Node` instances per JVM in tests. The historical versions can be static, as that doesn't change, but the featureservice itself needs to be instanced to the node to allow different nodes in the same JVM to have different features.

Reply:
I suppose there's no particular reason why we should limit the version for plugin historical features, if the plugin is updated and the feature is not...

Reply:
> as we need to be able to handle multiple Node instances per JVM in tests 

That's a very good reason